README.rst
About
=====

django_hits allow you to count hits to database objects and/or any unique identifier you want. It may
use multiple buckets for counting to keep different hits apart (list view of object is different from
detail page). In addition you may store historical data, if you want to (for example by calling "manage.py
hits_create_history" every week).


CHANGES.rst
Changelog
=========

0.4.0
-----

* Django 1.8 support
* Dropped support for Django < 1.6 (as we now use transaction.atomic)

0.3.0
-----

* Started CHANGES.rst

tests/settings.py
import warnings
warnings.simplefilter('always')

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': 'db.sqlite',
    },
}

USE_I18N = True
USE_L10N = True

INSTALLED_APPS = [
    'django.contrib.contenttypes',
    'django.contrib.auth',
    'django_hits',
    'tests',
]

MIDDLEWARE_CLASSES = ()

STATIC_URL = '/static/'

SECRET_KEY = '0'


tests/test_smoke.py
import pytest


def test_smoke():
    """Just some basic smoke tests (syntax, app loading)"""
    import django_hits
    import django_hits.models
    import django_hits.utils
    import django_hits.templatetags.hit_tags


def test_system_checks():
    import django

    if django.VERSION >= (1, 7, 0):
        from django.core import checks

        if django.VERSION >= (1, 8, 0):
            assert not checks.run_checks(include_deployment_checks=False), "Some Django system checks failed"
        else:
            assert not checks.run_checks(), "Some Django system checks failed"


@pytest.mark.django_db
def test_db():
    pass  # makes sure migrations are run


tests/test_basics.py
import pytest
from django_hits.models import Hit
from django.contrib.auth.models import AnonymousUser


@pytest.mark.django_db
def test_simple_hit():
    hit = Hit.objects.hit('something', None, '127.0.0.1')
    assert hit.views == 1
    assert hit.visits == 1
    hit = Hit.objects.hit('something', None, '127.0.0.1')
    assert hit.views == 2
    assert hit.visits == 1


django_hits/models.py
# -*- coding: utf-8 -*-
from django.db import models
from django.contrib.contenttypes import generic
from django.contrib.contenttypes.models import ContentType
from django.db import transaction, IntegrityError
from datetime import datetime, timedelta
from django.conf import settings


class HitManager(models.Manager):
    def get_for(self, obj, bucket=None):
        if bucket is None:
            bucket_kwargs = {'bucket__isnull': True}
        else:
            bucket_kwargs = {'bucket': bucket}
        if isinstance(obj, models.Model):
            content_type = ContentType.objects.get_for_model(obj.__class__)
            object_pk = getattr(obj, obj._meta.pk.column)
            try:
                return self.get_or_create(content_type=content_type, object_pk=object_pk, **bucket_kwargs)[0]
            except IntegrityError:  # catch race condition
                return self.get(content_type=content_type, object_pk=object_pk, **bucket_kwargs)
        elif isinstance(obj, (str, unicode)):
            try:
                return self.get_or_create(content_type__isnull=True, object_pk=obj, **bucket_kwargs)[0]
            except IntegrityError:  # catch race condition
                return self.get(content_type__isnull=True, object_pk=obj, **bucket_kwargs)
        else:
            raise Exception("Don't know what to do with this obj!?")

    def hit(self, obj, user, ip, bucket=None):
        hit = self.get_for(obj, bucket=bucket)
        hit.hit(user, ip)
        return hit


class Hit(models.Model):
    content_type = models.ForeignKey(ContentType, null=True)
    object_pk = models.CharField(max_length=50)  # TextField not possible, because unique_together is needed, must be enough
    content_object = generic.GenericForeignKey(ct_field="content_type", fk_field="object_pk")

    bucket = models.CharField(max_length=50, blank=True, null=True)  # Each object may have multiple buckets hits get counted in

    views = models.PositiveIntegerField(default=0)  # page hits/views
    visits = models.PositiveIntegerField(default=0)  # unique visits

    objects = HitManager()

    def hit(self, user, ip):
        try:
            with transaction.atomic():
                if self.has_hit_from(user, ip):
                    self.update_hit_from(user, ip)
                    Hit.objects.filter(pk=self.pk).update(views=models.F('views') + 1)
                    self.views += 1
                    return True
                else:
                    self.log.create(user=user, ip=ip)
                    Hit.objects.filter(pk=self.pk).update(views=models.F('views') + 1, visits=models.F('visits') + 1)
                    self.views += 1
                    self.visits += 1
                    return True
        except IntegrityError:
            # CATCH RACE CONDITION
            # log-extry was already created
            # happens when users double-click or reload to fast
            # (we ignore this)
            return False

    def has_hit_from(self, user, ip):
        self.clear_log()
        if self.log.filter(user=user, ip=ip).count():
            return True
        else:
            return False

    def update_hit_from(self, user, ip):
        self.log.filter(user=user, ip=ip).update(when=datetime.now())

    def clear_log(self):
        timespan = datetime.now() - timedelta(days=30)
        for l in self.log.filter(when__lt=timespan).order_by('-when')[25:]:
            l.delete()

    class Meta:
        unique_together = (('content_type', 'object_pk', 'bucket'),)


class HitLog(models.Model):
    hit = models.ForeignKey(Hit, related_name='log')
    user = models.ForeignKey(settings.AUTH_USER_MODEL, related_name='hits_log', null=True)
    ip = models.GenericIPAddressField(null=True)
    when = models.DateTimeField(default=datetime.now)

    class Meta:
        unique_together = (('hit', 'user', 'ip'),)


class HitHistory(models.Model):
    hit = models.ForeignKey(Hit, related_name='history')
    when = models.DateTimeField(default=datetime.now)

    views = models.PositiveIntegerField(default=0)
    visits = models.PositiveIntegerField(default=0)

    views_change = models.PositiveIntegerField(default=0)
    visits_change = models.PositiveIntegerField(default=0)





django_hits/templatetags/hit_tags.py
from django import template
from django_hits.models import Hit

register = template.Library()


class HitNode(template.Node):
    def __init__(self, context_var_name, as_var_name, bucket_var_name, count, count_bucket_only=False):
        self.context_var_name = context_var_name
        self.as_var_name = as_var_name
        self.bucket_var_name = bucket_var_name
        self.count = count
        self.count_bucket_only = count_bucket_only

    def _handle_hit(self, context, obj, bucket, count, user, ip):
        was_counted = False
        hit_cache_key = (obj, bucket)
        if hit_cache_key in context._hit_cache_:
            hit, was_counted = context._hit_cache_[hit_cache_key]
        else:
            hit = Hit.objects.get_for(obj, bucket=bucket)
        if count and not was_counted:
            was_counted = hit.hit(user, ip)
        context._hit_cache_[hit_cache_key] = (hit, was_counted)
        return hit


    def render(self, context):
        if not hasattr(context, '_hit_cache_'):
            context._hit_cache_ = {}
        bucket = None
        try:
            obj = self.context_var_name.resolve(context)
            if self.bucket_var_name:
                bucket = self.bucket_var_name.resolve(context)
        except template.VariableDoesNotExist:
            return ''
        try:
            user = template.Variable('user').resolve(context)
            if user.is_anonymous():
                user = None
        except template.VariableDoesNotExist:
            user = None
        count = self.count
        ip = None
        if count:
            try:
                request = template.Variable('request').resolve(context)
                if 'REMOTE_ADDR' in request.META:
                    ip = request.META['REMOTE_ADDR']
            except template.VariableDoesNotExist:
                pass
        hit = self._handle_hit(context, obj, bucket, count, user, ip)
        if not bucket is None:
            if not self.count_bucket_only:  # we count a global hit, too
                self._handle_hit(context, obj, None, count, user, ip)
        if self.as_var_name:
            context[self.as_var_name] = hit
        return ''


@register.tag
def get_hit(parser, token):
    '''
    {% get_hit for obj as hit %}
    {% get_hit for "static_page" as hit %}
    {% get_hit for obj bucket "foo" as hit %}
    {% get_hit for "static_page" bucket "foo" as hit %}
    '''
    tokens = token.split_contents()
    tag_name = tokens.pop(0)
    if not len(tokens) in (2, 4, 6,):
        raise template.TemplateSyntaxError, "%r tag requires 2, 4 or 6 arguments" % tag_name
    i = 0
    context_var_name = None
    as_var_name = None
    bucket_var_name = None
    while i < len(tokens):
        instruction = tokens[i]
        value = tokens[i + 1]
        if instruction == 'for':
            context_var_name = parser.compile_filter(value)
        elif instruction == 'as':
            as_var_name = value
        elif instruction == 'bucket':
            bucket_var_name = parser.compile_filter(value)
        else:
            raise template.TemplateSyntaxError, "Invalid instruction for %s tag ('%s')" % (tag_name, instruction)
        i += 2
    if context_var_name is None:
        raise template.TemplateSyntaxError, "%s tag requires some object to get hits for ('for')" % tag_name
    if as_var_name is None:
        raise template.TemplateSyntaxError, "%s tag requires some context name to write hit into ('as')" % tag_name
    return HitNode(context_var_name, as_var_name, bucket_var_name, False)


@register.tag
def count_hit(parser, token):
    '''
    {% count_hit for obj %}
    {% count_hit for obj as hit %}
    {% count_hit for "static_page" %}
    {% count_hit for "static_page" as hit %}
    {% count_hit for obj bucket "foo" %}
    {% count_hit for obj bucket "foo" as hit %}
    {% count_hit for "static_page" bucket "foo" %}
    {% count_hit for "static_page" bucket "foo" as hit %}
    {% count_hit for obj bucket "foo" only %}
    {% count_hit for obj bucket "foo" only as hit %}
    {% count_hit for "static_page" bucket "foo" only %}
    {% count_hit for "static_page" bucket "foo" only as hit %}
    '''
    tokens = token.split_contents()
    tag_name = tokens.pop(0)
    i = 0
    context_var_name = None
    as_var_name = None
    bucket_var_name = None
    count_bucket_only = False
    while i < len(tokens):
        try:
            instruction = tokens[i]
            value = tokens[i + 1]
        except IndexError:
            raise template.TemplateSyntaxError, "Invalid usage of %s tag" % tag_name
        if instruction == 'for':
            context_var_name = parser.compile_filter(value)
        elif instruction == 'as':
            as_var_name = value
        elif instruction == 'bucket':
            bucket_var_name = parser.compile_filter(value)
            try:
                _only = tokens[i + 2]
                if _only == 'only':
                    count_bucket_only = True
                    i += 1  # we have to eat one more token
            except IndexError:
                pass  # no problem, only is optional
        else:
            raise template.TemplateSyntaxError, "Invalid instruction for %s tag ('%s')" % (tag_name, instruction)
        i += 2
    if context_var_name is None:
        raise template.TemplateSyntaxError, "%s tag requires some object to count hits for ('for')" % tag_name
    return HitNode(context_var_name, as_var_name, bucket_var_name, True, count_bucket_only)



django_hits/migrations/0002_add_bucket.py
# -*- coding: utf-8 -*-
from south.utils import datetime_utils as datetime
from south.db import db
from south.v2 import SchemaMigration
from django.db import models


class Migration(SchemaMigration):

    def forwards(self, orm):
        # Removing unique constraint on 'Hit', fields ['content_type', 'object_pk']
        db.delete_unique(u'django_hits_hit', ['content_type_id', 'object_pk'])

        # Adding field 'Hit.bucket'
        db.add_column(u'django_hits_hit', 'bucket',
                      self.gf('django.db.models.fields.CharField')(max_length=50, null=True, blank=True),
                      keep_default=False)


        # Changing field 'Hit.object_pk'
        db.alter_column(u'django_hits_hit', 'object_pk', self.gf('django.db.models.fields.CharField')(max_length=50))
        # Adding unique constraint on 'Hit', fields ['content_type', 'object_pk', 'bucket']
        db.create_unique(u'django_hits_hit', ['content_type_id', 'object_pk', 'bucket'])


    def backwards(self, orm):
        # Removing unique constraint on 'Hit', fields ['content_type', 'object_pk', 'bucket']
        db.delete_unique(u'django_hits_hit', ['content_type_id', 'object_pk', 'bucket'])

        # Deleting field 'Hit.bucket'
        db.delete_column(u'django_hits_hit', 'bucket')


        # Changing field 'Hit.object_pk'
        db.alter_column(u'django_hits_hit', 'object_pk', self.gf('django.db.models.fields.CharField')(max_length=255))
        # Adding unique constraint on 'Hit', fields ['content_type', 'object_pk']
        db.create_unique(u'django_hits_hit', ['content_type_id', 'object_pk'])


    models = {
        u'auth.group': {
            'Meta': {'object_name': 'Group'},
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'name': ('django.db.models.fields.CharField', [], {'unique': 'True', 'max_length': '80'}),
            'permissions': ('django.db.models.fields.related.ManyToManyField', [], {'to': u"orm['auth.Permission']", 'symmetrical': 'False', 'blank': 'True'})
        },
        u'auth.permission': {
            'Meta': {'ordering': "(u'content_type__app_label', u'content_type__model', u'codename')", 'unique_together': "((u'content_type', u'codename'),)", 'object_name': 'Permission'},
            'codename': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'content_type': ('django.db.models.fields.related.ForeignKey', [], {'to': u"orm['contenttypes.ContentType']"}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '50'})
        },
        u'auth.user': {
            'Meta': {'object_name': 'User'},
            'date_joined': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            'email': ('django.db.models.fields.EmailField', [], {'max_length': '75', 'blank': 'True'}),
            'first_name': ('django.db.models.fields.CharField', [], {'max_length': '30', 'blank': 'True'}),
            'groups': ('django.db.models.fields.related.ManyToManyField', [], {'symmetrical': 'False', 'related_name': "u'user_set'", 'blank': 'True', 'to': u"orm['auth.Group']"}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'is_active': ('django.db.models.fields.BooleanField', [], {'default': 'True'}),
            'is_staff': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
            'is_superuser': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
            'last_login': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            'last_name': ('django.db.models.fields.CharField', [], {'max_length': '30', 'blank': 'True'}),
            'password': ('django.db.models.fields.CharField', [], {'max_length': '128'}),
            'user_permissions': ('django.db.models.fields.related.ManyToManyField', [], {'symmetrical': 'False', 'related_name': "u'user_set'", 'blank': 'True', 'to': u"orm['auth.Permission']"}),
            'username': ('django.db.models.fields.CharField', [], {'unique': 'True', 'max_length': '30'})
        },
        u'contenttypes.contenttype': {
            'Meta': {'ordering': "('name',)", 'unique_together': "(('app_label', 'model'),)", 'object_name': 'ContentType', 'db_table': "'django_content_type'"},
            'app_label': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'model': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '100'})
        },
        u'django_hits.hit': {
            'Meta': {'unique_together': "(('content_type', 'object_pk', 'bucket'),)", 'object_name': 'Hit'},
            'bucket': ('django.db.models.fields.CharField', [], {'max_length': '50', 'null': 'True', 'blank': 'True'}),
            'content_type': ('django.db.models.fields.related.ForeignKey', [], {'to': u"orm['contenttypes.ContentType']", 'null': 'True'}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'object_pk': ('django.db.models.fields.CharField', [], {'max_length': '50'}),
            'views': ('django.db.models.fields.PositiveIntegerField', [], {'default': '0'}),
            'visits': ('django.db.models.fields.PositiveIntegerField', [], {'default': '0'})
        },
        u'django_hits.hitlog': {
            'Meta': {'unique_together': "(('hit', 'user', 'ip'),)", 'object_name': 'HitLog'},
            'hit': ('django.db.models.fields.related.ForeignKey', [], {'related_name': "'log'", 'to': u"orm['django_hits.Hit']"}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'ip': ('django.db.models.fields.IPAddressField', [], {'max_length': '15', 'null': 'True'}),
            'user': ('django.db.models.fields.related.ForeignKey', [], {'related_name': "'hits_log'", 'null': 'True', 'to': u"orm['auth.User']"}),
            'when': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'})
        }
    }

    complete_apps = ['django_hits']

django_hits/management/commands/hits_create_history.py
from django.core.management.base import BaseCommand
from optparse import make_option


class Command(BaseCommand):
    option_list = BaseCommand.option_list + (
        make_option('-b', '--bucket', action='store', dest='bucket',
                    help='Select desired buckets to create history for (by default no bucket will be created)'),
        make_option('-a', '--all-buckets', action='store_true', dest='all_buckets',
                    help='Create history for all buckets'),
    )
    help = ''
    args = ''

    def handle(self, *filenames, **options):
        from ...models import Hit
        from ...utils import create_hit_history

        queryset = Hit.objects.all()
        if not options['all_buckets']:
            if options['bucket'] is None:
                queryset = queryset.filter(bucket__isnull=True)
            else:
                queryset = queryset.filter(bucket=options['bucket'])
        for hit in queryset:
            create_hit_history(hit)


django_hits/migrations/0004_auto__chg_field_hitlog_ip.py
# -*- coding: utf-8 -*-
from south.utils import datetime_utils as datetime
from south.db import db
from south.v2 import SchemaMigration
from django.db import models


class Migration(SchemaMigration):

    def forwards(self, orm):

        # Changing field 'HitLog.ip'
        db.alter_column(u'django_hits_hitlog', 'ip', self.gf('django.db.models.fields.GenericIPAddressField')(max_length=39, null=True))

    def backwards(self, orm):

        # Changing field 'HitLog.ip'
        db.alter_column(u'django_hits_hitlog', 'ip', self.gf('django.db.models.fields.IPAddressField')(max_length=15, null=True))

    models = {
        u'auth.group': {
            'Meta': {'object_name': 'Group'},
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'name': ('django.db.models.fields.CharField', [], {'unique': 'True', 'max_length': '80'}),
            'permissions': ('django.db.models.fields.related.ManyToManyField', [], {'to': u"orm['auth.Permission']", 'symmetrical': 'False', 'blank': 'True'})
        },
        u'auth.permission': {
            'Meta': {'ordering': "(u'content_type__app_label', u'content_type__model', u'codename')", 'unique_together': "((u'content_type', u'codename'),)", 'object_name': 'Permission'},
            'codename': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'content_type': ('django.db.models.fields.related.ForeignKey', [], {'to': u"orm['contenttypes.ContentType']"}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '50'})
        },
        u'auth.user': {
            'Meta': {'object_name': 'User'},
            'date_joined': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            'email': ('django.db.models.fields.EmailField', [], {'max_length': '75', 'blank': 'True'}),
            'first_name': ('django.db.models.fields.CharField', [], {'max_length': '30', 'blank': 'True'}),
            'groups': ('django.db.models.fields.related.ManyToManyField', [], {'symmetrical': 'False', 'related_name': "u'user_set'", 'blank': 'True', 'to': u"orm['auth.Group']"}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'is_active': ('django.db.models.fields.BooleanField', [], {'default': 'True'}),
            'is_staff': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
            'is_superuser': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
            'last_login': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            'last_name': ('django.db.models.fields.CharField', [], {'max_length': '30', 'blank': 'True'}),
            'password': ('django.db.models.fields.CharField', [], {'max_length': '128'}),
            'user_permissions': ('django.db.models.fields.related.ManyToManyField', [], {'symmetrical': 'False', 'related_name': "u'user_set'", 'blank': 'True', 'to': u"orm['auth.Permission']"}),
            'username': ('django.db.models.fields.CharField', [], {'unique': 'True', 'max_length': '30'})
        },
        u'contenttypes.contenttype': {
            'Meta': {'ordering': "('name',)", 'unique_together': "(('app_label', 'model'),)", 'object_name': 'ContentType', 'db_table': "'django_content_type'"},
            'app_label': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'model': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '100'})
        },
        u'django_hits.hit': {
            'Meta': {'unique_together': "(('content_type', 'object_pk', 'bucket'),)", 'object_name': 'Hit'},
            'bucket': ('django.db.models.fields.CharField', [], {'max_length': '50', 'null': 'True', 'blank': 'True'}),
            'content_type': ('django.db.models.fields.related.ForeignKey', [], {'to': u"orm['contenttypes.ContentType']", 'null': 'True'}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'object_pk': ('django.db.models.fields.CharField', [], {'max_length': '50'}),
            'views': ('django.db.models.fields.PositiveIntegerField', [], {'default': '0'}),
            'visits': ('django.db.models.fields.PositiveIntegerField', [], {'default': '0'})
        },
        u'django_hits.hithistory': {
            'Meta': {'object_name': 'HitHistory'},
            'hit': ('django.db.models.fields.related.ForeignKey', [], {'related_name': "'history'", 'to': u"orm['django_hits.Hit']"}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'views': ('django.db.models.fields.PositiveIntegerField', [], {'default': '0'}),
            'views_change': ('django.db.models.fields.PositiveIntegerField', [], {'default': '0'}),
            'visits': ('django.db.models.fields.PositiveIntegerField', [], {'default': '0'}),
            'visits_change': ('django.db.models.fields.PositiveIntegerField', [], {'default': '0'}),
            'when': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'})
        },
        u'django_hits.hitlog': {
            'Meta': {'unique_together': "(('hit', 'user', 'ip'),)", 'object_name': 'HitLog'},
            'hit': ('django.db.models.fields.related.ForeignKey', [], {'related_name': "'log'", 'to': u"orm['django_hits.Hit']"}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'ip': ('django.db.models.fields.GenericIPAddressField', [], {'max_length': '39', 'null': 'True'}),
            'user': ('django.db.models.fields.related.ForeignKey', [], {'related_name': "'hits_log'", 'null': 'True', 'to': u"orm['auth.User']"}),
            'when': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'})
        }
    }

    complete_apps = ['django_hits']

django_hits/migrations/0003_added_history.py
# -*- coding: utf-8 -*-
from south.utils import datetime_utils as datetime
from south.db import db
from south.v2 import SchemaMigration
from django.db import models


class Migration(SchemaMigration):

    def forwards(self, orm):
        # Adding model 'HitHistory'
        db.create_table(u'django_hits_hithistory', (
            (u'id', self.gf('django.db.models.fields.AutoField')(primary_key=True)),
            ('hit', self.gf('django.db.models.fields.related.ForeignKey')(related_name='history', to=orm['django_hits.Hit'])),
            ('when', self.gf('django.db.models.fields.DateTimeField')(default=datetime.datetime.now)),
            ('views', self.gf('django.db.models.fields.PositiveIntegerField')(default=0)),
            ('visits', self.gf('django.db.models.fields.PositiveIntegerField')(default=0)),
            ('views_change', self.gf('django.db.models.fields.PositiveIntegerField')(default=0)),
            ('visits_change', self.gf('django.db.models.fields.PositiveIntegerField')(default=0)),
        ))
        db.send_create_signal(u'django_hits', ['HitHistory'])


    def backwards(self, orm):
        # Deleting model 'HitHistory'
        db.delete_table(u'django_hits_hithistory')


    models = {
        u'auth.group': {
            'Meta': {'object_name': 'Group'},
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'name': ('django.db.models.fields.CharField', [], {'unique': 'True', 'max_length': '80'}),
            'permissions': ('django.db.models.fields.related.ManyToManyField', [], {'to': u"orm['auth.Permission']", 'symmetrical': 'False', 'blank': 'True'})
        },
        u'auth.permission': {
            'Meta': {'ordering': "(u'content_type__app_label', u'content_type__model', u'codename')", 'unique_together': "((u'content_type', u'codename'),)", 'object_name': 'Permission'},
            'codename': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'content_type': ('django.db.models.fields.related.ForeignKey', [], {'to': u"orm['contenttypes.ContentType']"}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '50'})
        },
        u'auth.user': {
            'Meta': {'object_name': 'User'},
            'date_joined': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            'email': ('django.db.models.fields.EmailField', [], {'max_length': '75', 'blank': 'True'}),
            'first_name': ('django.db.models.fields.CharField', [], {'max_length': '30', 'blank': 'True'}),
            'groups': ('django.db.models.fields.related.ManyToManyField', [], {'symmetrical': 'False', 'related_name': "u'user_set'", 'blank': 'True', 'to': u"orm['auth.Group']"}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'is_active': ('django.db.models.fields.BooleanField', [], {'default': 'True'}),
            'is_staff': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
            'is_superuser': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
            'last_login': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            'last_name': ('django.db.models.fields.CharField', [], {'max_length': '30', 'blank': 'True'}),
            'password': ('django.db.models.fields.CharField', [], {'max_length': '128'}),
            'user_permissions': ('django.db.models.fields.related.ManyToManyField', [], {'symmetrical': 'False', 'related_name': "u'user_set'", 'blank': 'True', 'to': u"orm['auth.Permission']"}),
            'username': ('django.db.models.fields.CharField', [], {'unique': 'True', 'max_length': '30'})
        },
        u'contenttypes.contenttype': {
            'Meta': {'ordering': "('name',)", 'unique_together': "(('app_label', 'model'),)", 'object_name': 'ContentType', 'db_table': "'django_content_type'"},
            'app_label': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'model': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '100'})
        },
        u'django_hits.hit': {
            'Meta': {'unique_together': "(('content_type', 'object_pk', 'bucket'),)", 'object_name': 'Hit'},
            'bucket': ('django.db.models.fields.CharField', [], {'max_length': '50', 'null': 'True', 'blank': 'True'}),
            'content_type': ('django.db.models.fields.related.ForeignKey', [], {'to': u"orm['contenttypes.ContentType']", 'null': 'True'}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'object_pk': ('django.db.models.fields.CharField', [], {'max_length': '50'}),
            'views': ('django.db.models.fields.PositiveIntegerField', [], {'default': '0'}),
            'visits': ('django.db.models.fields.PositiveIntegerField', [], {'default': '0'})
        },
        u'django_hits.hithistory': {
            'Meta': {'object_name': 'HitHistory'},
            'hit': ('django.db.models.fields.related.ForeignKey', [], {'related_name': "'history'", 'to': u"orm['django_hits.Hit']"}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'views': ('django.db.models.fields.PositiveIntegerField', [], {'default': '0'}),
            'views_change': ('django.db.models.fields.PositiveIntegerField', [], {'default': '0'}),
            'visits': ('django.db.models.fields.PositiveIntegerField', [], {'default': '0'}),
            'visits_change': ('django.db.models.fields.PositiveIntegerField', [], {'default': '0'}),
            'when': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'})
        },
        u'django_hits.hitlog': {
            'Meta': {'unique_together': "(('hit', 'user', 'ip'),)", 'object_name': 'HitLog'},
            'hit': ('django.db.models.fields.related.ForeignKey', [], {'related_name': "'log'", 'to': u"orm['django_hits.Hit']"}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'ip': ('django.db.models.fields.IPAddressField', [], {'max_length': '15', 'null': 'True'}),
            'user': ('django.db.models.fields.related.ForeignKey', [], {'related_name': "'hits_log'", 'null': 'True', 'to': u"orm['auth.User']"}),
            'when': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'})
        }
    }

    complete_apps = ['django_hits']